### ISSUE TITLE
Remove openlibm

### ISSUE THREAD

**Issue by simonbyrne** [JuliaLang/julia#26434]
Thanks to @pkofod we're quite close, but there are a couple of places left:
- [ ] [fastmath.jl](https://github.com/JuliaLang/julia/blob/master/base/fastmath.jl) still calls the `libm` in a few places
- [ ] [`cbrt`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/math.jl#L247-L253) #25948
- [ ] [`exp2`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/math.jl#L247-L253) #24859
- [ ] [`expm1`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/math.jl#L247-L253) #25903
- [ ] [`log2`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/math.jl#L431-L432)
- [ ] [`log10`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/math.jl#L431-L432)
- [ ] [`lgamma`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/math.jl#L431-L432)
- [ ] [`lgamma_r`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/special/gamma.jl#L13-L22)
- [ ] [`gamma`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/special/gamma.jl#L3-L4)
- [ ] [`rem`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/math.jl#L679-L682)
- [ ] [`modf`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/math.jl#L700-L710)
- [ ] [`fma`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/floatfuncs.jl#L278-L298) to deal with broken system libms (#9847)
- [ ] [`_ldexp_exp`](https://github.com/JuliaLang/julia/blob/46dcb35e1b92bf252aee078a90e42557b48376bc/base/special/hyperbolic.jl#L15-L16)

**Comment by vchuravy** [comment]
Aren't we also convincing LLVM to call Openlibm instead of the system libm?

**Comment by simonbyrne** [comment]
Are we? I remember talking about it, but did we ever make that happen?

**Comment by vchuravy** [comment]
I can't find any reference to it in the source code so we are probably not doing that and I just remember the same discussion.
We could probably resolve the libcalls llvm will lower them to similar to what I am trying to do in #26381

**Comment by simonbyrne** [comment]
Could we get them to resolve to specific Julia methods? That would be ideal.

**Comment by Keno** [comment]
Ideally we should figure out a generic mechanism to have LLVM request a function be constant folded with the implementation we have available in the compiler.

**Comment by vchuravy** [comment]
Speculating here, but it looks like it might be possible to use
`setAvailableWithName` after using `getLibFunc`
http://llvm.org/doxygen/TargetLibraryInfo_8h_source.html to force a
particular triple to use a different version of a library function (and it
might even allow us to insert vectorised versions?)

**Comment by pkofod** [comment]
I have code for `log2`, `log10`, `lgamma` and `remainder` as well. I think `_ldexp_exp` as well, but it is very short either way. I just havn't really prioritized making new PRs given that the existing ones are moving rather slowly, and I have some deadlines outside of OSS.

**Comment by ararslan** [comment]
@pkofod If you get the chance I recommend just opening PRs with your changes, and if you need to be away for a while, someone who knows what they're doing can push to the branch with any necessary changes if applicable.

**Comment by giordano** [comment]
Regarding fastmath point, I have this PR still waiting for review: https://github.com/JuliaLang/julia/pull/24031.  If it's good, I can add the missing functions

**Comment by ViralBShah** [comment]
@pkofod Would you be able to push through your PRs?

Things like SpecialFunctions should be easy to update, and the fastmath could just fallback to the system libm initially perhaps?

**Comment by pkofod** [comment]
What do you mean? If I can pr the remaining, or if the existing can be be finished soon ?

**Comment by ViralBShah** [comment]
I was actually referring to bumping the existing ones if someone needs to review or do something about those. How far do you think we are from excising the remaining ones?

**Comment by StefanKarpinski** [comment]
The above checklist with no check marks makes it look like none of this is done which I'm fairly sure it not the case at this point. @pkofod, would you mind updating the checkmarks here? Should be very satisfying since everyone knows that getting to check the boxes is why people make check lists.

**Comment by pkofod** [comment]
Well, I think it might actually be quite accurate, as this was made after most of my prs were merged. The three open PRs are quite done as far as I remember, but I'm just waiting for feedback. I guess they have to be rebased by now, but I'd be happy to help get them merged asap to the extent that someone agrees to review. I'm on a quite tight time budget these months, so I can really only justify spending time on something that has a high chance of moving forward.

**Comment by ViralBShah** [comment]
Let's tag folks who could help review and move things along. Perhaps @simonbyrne ?

**Comment by pkofod** [comment]
I’ll rebase the existing PRs so we have something to work off of

**Comment by pkofod** [comment]
I can't edit @simonbyrne 's post, but https://github.com/JuliaLang/julia/pull/27137 covers log2 and log10

**Comment by einzigsue** [comment]
> I have code for `log2`, `log10`, `lgamma` and `remainder` as well. I think `_ldexp_exp` as well, but it is very short either way. I just havn't really prioritized making new PRs given that the existing ones are moving rather slowly, and I have some deadlines outside of OSS.

I found these two lines in the file base/special/hyperbolic.jl
_ldexp_exp(x::Float64, i::T) where T = ccall(("__ldexp_exp", libm), Float64, (Float64, T), x, i)
_ldexp_exp(x::Float32, i::T) where T = ccall(("__ldexp_expf",libm), Float32, (Float32, T), x, i)
but neither gnu libm.so.6 nor intel/2018.2.199 libimf.so has the functions __ldexp_exp and __ldexp_expf defined.
 
Does it mean v1.0.0 has to be built with openlibm?

**Comment by ararslan** [comment]
Correct. Julia's build system builds and bundles openlibm. It's largely invisible to the user.

**Comment by einzigsue** [comment]
then it might be better to disable USE_INTEL_LIBM from Make.inc

**Comment by ararslan** [comment]
Sorry, I misread. No it doesn't _have_ to be built with openlibm, but we strongly encourage it. Every `USE_SYSTEM_*` flag for Make is effectively "use at your own risk."

**Comment by einzigsue** [comment]
I thought USE_INTEL_LIBM is to configure intel tools.

**Comment by pkofod** [comment]
@einzigsue ah yes I can see why that would be an issue if you try to swap out the const `libm` with some other libm that otherwise have all the functions we are calling using call. That’ll be a priority to get out.

**Comment by pkofod** [comment]
__ldexp_exp can be crossed off the list

**Comment by ararslan** [comment]
> SpecialFunctions.jl also calls a few

By the magic of BinaryBuilder/BinaryProvider, we can have SpecialFunctions pull in openlibm as needed without needing to assume it lives in Julia.

**Comment by ViralBShah** [comment]
I've updated this issue since a lot more work has been done. I think we can use the LLVM provided or system provided libm for the stuff that is in base, since we only use libm for only `remainder` and `modf`. 

Probably, we can replace the use of those altogether - but we may be ready to remove openlibm altogether.

SpecialFunctions.jl can then have a direct dependency on `openlibm` through BB, which should be straightforward. 

cc @oscardssmith who may know most about the status of these.

**Comment by oscardssmith** [comment]
I believe your summary is about right. `_ldexp_exp` no longer exists in `Base` though. I believe the old `exp` implementations needed it, but it was removed from 1.6. `rem` and `modf` should be pretty easy to port. `fma` will be harder. I believe that this list is currently missing `pow`, which I am planning on replacing soon.

**Comment by ViralBShah** [comment]
`fma` only uses openlibm on 32-bit I thought.

**Comment by oscardssmith** [comment]
What about `arm` architectures that don't have native fma instructions?

**Comment by ViralBShah** [comment]
Doesn't llvm do something there, like fallback to the non-fma version or something? I mean having openlibm does not solve that problem anyways, right?

**Comment by oscardssmith** [comment]
I'm not the right person to ask here. I've done pretty much all my work assuming 64 bit x86...

**Comment by ViralBShah** [comment]
I'll submit a PR that removes openlibm, and we'll see where it goes.

**Comment by ViralBShah** [comment]
As seen in #42299 Linux32 seems to have a lot of failures when not using openlibm. It seems that we should move `remainder`, `modff` and `fma` to avoid openlibm before we can excise it.

**Comment by KristofferC** [comment]
The fma bug that is linked in https://github.com/JuliaLang/julia/issues/9847 was fixed in 2014: https://sourceware.org/bugzilla/show_bug.cgi?id=16064. I wonder if it is still relevant.

**Comment by ViralBShah** [comment]
Can we count on the system libm and llvm to provide `fma`? https://github.com/JuliaLang/julia/blob/1b8cf40dc0936d657222fe44a4993a580d63098e/base/floatfuncs.jl#L358

**Comment by oscardssmith** [comment]
I'll take a look at a native Julia version by next weekend.  If we want Julia to do it, we do need to figure out how to do the versioning.

**Comment by ararslan** [comment]
I started on a native implementation of `fma` based on msun but haven't got it finished, cleaned up, and/or properly tested. Don't let that stop you though, Oscar, if you wanted to have a go at it.

**Comment by StefanKarpinski** [comment]
Can we emulate FMA with compensated arithmetic?

**Comment by simonbyrne** [comment]
Yes, but FMAs are notoriously difficult to get right for all the edge cases. A simpler idea is to just call the `fma` function, and manually reset the rounding mode each time (#42961): since this is just to deal with very old 32-bit linux installations, it seems like the simplest option.

**Comment by KristofferC** [comment]
I thought we already merged FMA though: https://github.com/JuliaLang/julia/pull/42783?

**Comment by oscardssmith** [comment]
yeah. it is a total pain to get right, but we now have it in Base.

**Comment by simonbyrne** [comment]
Oh, or we could do that...

**Comment by PallHaraldsson** [comment]
>fma is the last issue to remove openlibm from base. https://github.com/JuliaLang/julia/pull/42299#issuecomment-944932097

>I thought we already merged FMA though: https://github.com/JuliaLang/julia/pull/42783?

Can't we just remove openlibm, and ditch those users, without fma hardware? Or use the emulated one?

>FMA3 is supported in AMD processors starting with the [Piledriver](https://en.wikipedia.org/wiki/Piledriver_(microarchitecture)) architecture and [Intel](https://en.wikipedia.org/wiki/Intel_Corporation) starting with [Haswell processors](https://en.wikipedia.org/wiki/Haswell_(microarchitecture)) and [Broadwell processors](https://en.wikipedia.org/wiki/Broadwell_(microarchitecture)) since 2014.

I think it may be actually CPUs older than June 2013, so 9+ years old.

**Comment by StefanKarpinski** [comment]
It kind of looks like Base Julia doesn't need to depend on OpenLibm anymore? At least based on the checkmark list at the top. Is that the case now? (@oscardssmith, @simonbyrne)

**Comment by oscardssmith** [comment]
See https://github.com/JuliaLang/julia/pull/42299
